components/omega/doc/design/EOS.md

```diff
@@ -0,0 +1,158 @@
+
+### 2.7 Desired: output in situ temperature
+Later versions should include calculating the in situ  temperature needed for
+coupling (at surface including at non-zero pressure).
+
+### 2.8 Desired: output first derivatives
+The EOS will need to be able to produce alpha and beta (drho/dT and drho/dS)
+for linear expansions in the higher-order pressure gradient.
+
+### 2.9 Desired: check value range
+Later versions should include a check on the range of ocean properties to
+assess use of TEOS-10 (equivalent to the ``ocean funnel'' check).
+
+### 2.10 Desired: check value range
+Later versions should include functions to convert between conservative and
+potential temperatures, and absolute and practical salinity. These will be used
+to convert the initial conditions or compare to ocean states from other sources
+(MPAS-Ocean, reanalysis, other models etc.) These functions may be used offline
+in pre- or post-processing but need to be consistent with the EOS
+implementation. 
+
+## 3 Algorithmic Formulation
+
+## 4 Design
+The GSW-C toolbox will be incorporated into Omega as a submodule.
+In order to abide by the license of the toolbox, the toolbox will only be used
+in testing to check our implementation. The coefficients for the Roquet et al.
+2015 expansion will be based on the published values (Appendix). A Kokkos
+implementation of the function to compute specific volume will be ported to
+Omega. The GSW-C toolbox sub-module will serve as a baseline reference for our
+ports in unit tests. 
+
+For flexibility in optimizing performance, calls to the EOS function should be
+able to support both fine-grained single cell and whole array computation.
+Given the limitations with using virtual functions for inheritance on the
+device, EOS functions will be provided as functors with a consistent interface.
+Functions that use the EOS (e.g. pressure gradient) will have a template
+argument for the EOS type. A runtime switch statement is then used to call the
+user function with the EOS selected in the configuration file.  This approach
+allows the EOS to be called within parallel loops without any inner logic
+statements and maintains the flexibility to calculate the EOS for a single cell
+so EOS calculations can be fused with other computation.  If performance can be
+improved by pre-calculating the EOS, the functors can be used to provide
+whole-array calculations functions as well.
+
+### 4.1 Data types and parameters
+An `enum class` will be used to specify options for the EOS used for an Omega
+simulation:
+```c++
+enum class EOSType{
+   Linear,  /// Linear equation of state
+   TEOS10Poly75t.  /// Roquet et al. 2015 75 term expansion
+}
+```
+The user will select the EOS type at runtime in the input yaml file under an
+eos section:
+```yaml
+    eos:
+       eosType: 'teos10'
+```
+### 4.2 Functions
+Both the linear and TEOS-10 EOS options will have a functor that provides the
+same interface:
+```c++
+class LinearEOS {
```

Comment:
Do we really need a separate class for each EOS choice? I think we should have just a single EOS class (see related comment later)